sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetUtils.scala

```diff
@@ -127,4 +147,360 @@ object ParquetUtils {
+   *         and the second element is the aggregated value.
+   */
+  private[sql] def getPushedDownAggResult(
+      footer: ParquetMetadata,
+      dataSchema: StructType,
+      aggregation: Aggregation)
+  : (Array[PrimitiveType.PrimitiveTypeName], Array[Any]) = {
+    val footerFileMetaData = footer.getFileMetaData
+    val fields = footerFileMetaData.getSchema.getFields
+    val blocks = footer.getBlocks()
+    val typesBuilder = ArrayBuilder.make[PrimitiveType.PrimitiveTypeName]
+    val valuesBuilder = ArrayBuilder.make[Any]
+
+    for (i <- 0 until aggregation.aggregateExpressions.size) {
+      var value: Any = None
+      var rowCount = 0L
+      var isCount = false
+      var index = 0
+      blocks.forEach { block =>
+        val blockMetaData = block.getColumns()
+        aggregation.aggregateExpressions(i) match {
+          case Max(col, _) =>
+            index = dataSchema.fieldNames.toList.indexOf(col)
+            val currentMax = getCurrentBlockMaxOrMin(footer, blockMetaData, index, true)
+            if (currentMax != None &&
+              (value == None || currentMax.asInstanceOf[Comparable[Any]].compareTo(value) > 0)) {
+              value = currentMax
+            }
+
+          case Min(col, _) =>
+            index = dataSchema.fieldNames.toList.indexOf(col)
+            val currentMin = getCurrentBlockMaxOrMin(footer, blockMetaData, index, false)
+            if (currentMin != None &&
+              (value == None || currentMin.asInstanceOf[Comparable[Any]].compareTo(value) < 0)) {
+              value = currentMin
+            }
+
+          case Count(col, _, _) =>
+            index = dataSchema.fieldNames.toList.indexOf(col)
+            rowCount += getRowCountFromParquetMetadata(footer)
+            if (!col.equals("1")) {  // "1" is for count(*)
+              rowCount -= getNumNulls(footer, blockMetaData, index)
+            }
+            isCount = true
+
+          case _ =>
+        }
+      }
+      if (isCount) {
+        valuesBuilder += rowCount
+        typesBuilder += PrimitiveType.PrimitiveTypeName.INT96
+      } else {
+        valuesBuilder += value
+        typesBuilder += fields.get(index).asPrimitiveType.getPrimitiveTypeName
+      }
+    }
+    (typesBuilder.result(), valuesBuilder.result())
+  }
+
+  /**
+   * get the Max or Min value for ith column in the current block
+   *
+   * @return the Max or Min value
+   */
+  private def getCurrentBlockMaxOrMin(
+      footer: ParquetMetadata,
+      columnChunkMetaData: util.List[ColumnChunkMetaData],
+      i: Int,
+      isMax: Boolean): Any = {
+    val parquetType = footer.getFileMetaData.getSchema.getType(i)
+    if (!parquetType.isPrimitive) {
+      throw new IllegalArgumentException("Unsupported type : " + parquetType.toString)
+    }
+    val statistics = columnChunkMetaData.get(i).getStatistics()
+    if (isMax) statistics.genericGetMax() else statistics.genericGetMin()
```

Comment:
just want to double check, Parquet will always make sure the min/max statistics to be presented in footer, right?

Reply:
Good question. I actually need to check if Parquet returns the min/max statistics. If not, I will either throw Exception or fall back to the no push down way. I think fall back is a better solution.

Reply:
I can't find a good way to fall back. We won't be able to read footer until `FilePartitionReaderFactory.createReader`, that's when we can get a partition of the file to read. Seems to me it's too late to fall back at that time. I looked at Presto parquet partial aggregation implementation, and it throws Exception. I will throw Exception for now. If anybody has a better idea, please let know.